linode/profile/framework_datasource.go

```diff
@@ -0,0 +1,148 @@
+)
+
+func NewDataSource() datasource.DataSource {
+	return &DataSource{}
+}
+
+type DataSource struct {
+	client *linodego.Client
+}
+
+func (data *DataSourceModel) parseProfile(ctx context.Context, profile *linodego.Profile) diag.Diagnostics {
+	data.Email = types.StringValue(profile.Email)
+	data.Timezone = types.StringValue(profile.Timezone)
+	data.EmailNotifications = types.BoolValue(profile.EmailNotifications)
+	data.Username = types.StringValue(profile.Username)
+	data.IPWhitelistEnabled = types.BoolValue(profile.IPWhitelistEnabled)
+	data.LishAuthMethod = types.StringValue(string(profile.LishAuthMethod))
+
+	authorized_keys, err := types.ListValueFrom(ctx, types.StringType, profile.AuthorizedKeys)
+	if err != nil {
+		return err
+	}
+
+	data.AuthorizedKeys = authorized_keys
+
+	data.TwoFactorAuth = types.BoolValue(profile.TwoFactorAuth)
+	data.Restricted = types.BoolValue(profile.Restricted)
+
+	referrals, err := flattenReferral(ctx, profile.Referrals)
+	if err != nil {
+		return err
+	}
+
+	data.Referrals = *referrals
+
+	id, _ := json.Marshal(profile)
+
+	data.ID = types.StringValue(string(id))
+
+	return nil
+}
+
+func (d *DataSource) Configure(
+	ctx context.Context,
+	req datasource.ConfigureRequest,
+	resp *datasource.ConfigureResponse,
+) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	meta := helper.GetDataSourceMeta(req, resp)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	d.client = meta.Client
+}
+
+type DataSourceModel struct {
+	Email              types.String `tfsdk:"email"`
+	Timezone           types.String `tfsdk:"timezone"`
+	EmailNotifications types.Bool   `tfsdk:"email_notifications"`
+	Username           types.String `tfsdk:"username"`
+	IPWhitelistEnabled types.Bool   `tfsdk:"ip_whitelist_enabled"`
+	LishAuthMethod     types.String `tfsdk:"lish_auth_method"`
+	AuthorizedKeys     types.List   `tfsdk:"authorized_keys"`
+	TwoFactorAuth      types.Bool   `tfsdk:"two_factor_auth"`
+	Restricted         types.Bool   `tfsdk:"restricted"`
+	Referrals          types.Object `tfsdk:"referrals"`
+	ID                 types.String `tfsdk:"id"`
+}
+
+func (d *DataSource) Metadata(
+	ctx context.Context,
+	req datasource.MetadataRequest,
+	resp *datasource.MetadataResponse,
+) {
+	resp.TypeName = "linode_profile"
+}
+
+func (d *DataSource) Schema(
+	ctx context.Context,
+	req datasource.SchemaRequest,
+	resp *datasource.SchemaResponse,
+) {
+	resp.Schema = frameworkDatasourceSchema
+}
+
+func (d *DataSource) Read(
+	ctx context.Context,
+	req datasource.ReadRequest,
+	resp *datasource.ReadResponse,
+) {
+	client := d.client
+
+	var data DataSourceModel
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	profile, err := client.GetProfile(ctx)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to get LKE Versions: %s", err.Error(),
+		)
+		return
+	}
+
+	data.parseProfile(ctx, profile)
```

Comment:
We should append the diagnostics returned from the parse function: 
```suggestion
	resp.Diagnostics.Append(data.parseProfile(ctx, profile))
	if resp.Diagnostics.HasError() {
		return
	}
```

Reply:
Fixed!